kobo/apps/openrosa/apps/logger/management/commands/clean_duplicated_submissions.py

```diff
@@ -39,99 +38,158 @@ def add_arguments(self, parser):
 
         if username:
             query = query.filter(xform__user__username=username)
 
-        query = query.values_list('uuid', flat=True)\
-            .annotate(count_uuid=Count('uuid'))\
-            .filter(count_uuid__gt=1)\
+        query = (
+            query.values_list('xml_hash', flat=True)
+            .annotate(count_xml_hash=Count('xml_hash'))
+            .filter(count_xml_hash__gt=1)
             .distinct()
+        )
 
-        for uuid in query.all():
-
-            duplicated_query = Instance.objects.filter(uuid=uuid)
-
-            instances_with_same_uuid = duplicated_query.values_list('id',
-                                                                    'xml_hash')\
-                .order_by('xml_hash', 'date_created')
-            xml_hash_ref = None
-            instance_id_ref = None
-
-            duplicated_instance_ids = []
-            for instance_with_same_uuid in instances_with_same_uuid:
-                instance_id = instance_with_same_uuid[0]
-                instance_xml_hash = instance_with_same_uuid[1]
-
-                if instance_xml_hash != xml_hash_ref:
-                    self.__clean_up(instance_id_ref,
-                                    duplicated_instance_ids)
-                    xml_hash_ref = instance_xml_hash
-                    instance_id_ref = instance_id
-                    duplicated_instance_ids = []
-                    continue
-
-                duplicated_instance_ids.append(instance_id)
-
-            self.__clean_up(instance_id_ref,
-                            duplicated_instance_ids)
-
-        if not self.__vaccuum:
-            self.stdout.write('No instances have been purged.')
-        else:
-            # Update number of submissions for each user.
-            for user_ in list(self.__users):
-                result = XForm.objects.filter(user_id=user_.id)\
-                    .aggregate(count=Sum('num_of_submissions'))
-                user_.profile.num_of_submissions = result['count']
-                self.stdout.write(
-                    "\tUpdating `{}`'s number of submissions".format(
-                        user_.username))
-                user_.profile.save(update_fields=['num_of_submissions'])
+        for xml_hash in query.iterator():
+
+            duplicates_queryset = Instance.objects.filter(xml_hash=xml_hash)
+
+            instances_with_same_xml_hash = duplicates_queryset.values(
+                'id', 'uuid', 'xform_id'
+            ).order_by('xform_id', 'uuid', 'date_created')
+
+            duplicates_by_xform = self._get_duplicates_by_xform(
+                instances_with_same_xml_hash
+            )
+
+            for (
+                xform_id,
+                instances_with_same_xml_hash,
+            ) in duplicates_by_xform.items():
+                instance_ref = instances_with_same_xml_hash.pop(0)
+                self._clean_up(instance_ref, instances_with_same_xml_hash)
```

Comment:
Refactor the code to get all duplicates with same XML hash, and all duplicates with a different XML hash. 
All duplicates with the same XML hash should removed (and their attachments linked to `instance_ref`) and all duplicates with a different XML hash should have only their `uuid` replaced with the new string at line ...